File: verifier-ray/codegen/grandproduct.go
package codegen

import (
	"github.com/LFDT-Lineth/lineth-monorepo/prover-ray/wiop"
	"github.com/LFDT-Lineth/lineth-monorepo/prover-ray/wiop/compilers/grandproduct"
	"github.com/LFDT-Lineth/lineth-monorepo/prover-ray/wiop/compilers/messagebus"
)

// GrandProductSystem is the compiled metadata for every wiop.GrandProduct
// query in a wiop.System, in the form the Zig grandproduct sub-verifier
// consumes.
//
// The Z-recurrence and the row-0 boundary are ordinary vanishing constraints
// (registered by the grandproduct compiler's buildZ), so they are discharged
// by the vanishing sub-verifier. What remains for this sub-verifier is the
// pair of boundary identities the compiler's verifier actions enforce:
//
//   - FinalProductCheck (always present, one per GrandProduct):
//     ∏_entries Z[n-1] == Result
//   - a query-specific check on Result itself, whichever of
//     grandproduct.CheckResultIsOne or messagebus.CheckHandleSumInShard the
//     compiler attached to this GrandProduct (permutation arguments always
//     attach the former; the message-bus pass attaches the latter, skippable
//     per handle via MessageBus.SkipInShardCheck) — both reduce to
//     Result == Expected for a compile-time constant Expected.
//
// All cell operands are (round, index) references that the Zig verifier reads
// from ctx.rounds at verify time, so the check is against the adversary's
// transcript rather than baked-in honest-prover values.
type GrandProductSystem struct {
	SourceName string
	Queries    []GrandProductQuery
}

// GrandProductQuery is one reduced wiop.GrandProduct: the transcript
// positions of the Z-column endpoints and the claimed Result, plus the
// optional Result == Expected boundary check.
type GrandProductQuery struct {
	SourceName string
	// ZFinalRefs are the (round, index) positions of Z[n-1] for each Z column
	// FinalProductCheck folds into the claimed product.
	ZFinalRefs []ScalarCellRef
	// ResultRef is the (round, index) position of the claimed grand product.
	ResultRef ScalarCellRef
	// HasExpected reports whether a CheckResultIsOne / CheckHandleSumInShard
	// verifier action also constrains ResultRef to Expected. False for a
	// GrandProduct whose in-shard check was skipped (messagebus's
	// SkipInShardCheck, left to a downstream cross-shard layer).
	HasExpected bool
	// Expected is the base-field constant ResultRef must equal, valid only
	// when HasExpected is true. field.ElemOne() for both CheckResultIsOne and
	// the default (unsharded) CheckHandleSumInShard.
	Expected uint64
}

// BuildGrandProductSystem extracts every wiop.GrandProduct registered on sys
// and records its Z-endpoint / Result cell-reference coordinates, together
// with whichever Result == Expected check (if any) the compiler attached.
// Queries are collected in round/registration order so the output is
// deterministic.
//
// Requires the grandproduct compiler's FinalProductCheck to have been
// registered for every GrandProduct (compileGrandProducts runs unconditionally
// for every GrandProduct, including ones a caller — e.g. messagebus —
// constructs directly), and global.Compile to have been called afterwards so
// no cell lands in the last wiop round (protocol.replay excludes it from
// ctx.rounds).
func BuildGrandProductSystem(sys *wiop.System) (GrandProductSystem, error) {
	out := GrandProductSystem{SourceName: sys.Context.Path()}
	lastSlot := len(sys.Rounds) - 1

	expectedByResultCell := map[*wiop.Cell]uint64{}
	for _, round := range sys.Rounds {
		for _, action := range round.VerifierActions {
			switch a := action.(type) {
			case *grandproduct.CheckResultIsOne:
				expectedByResultCell[a.GrandProduct.Result] = 1
			case *messagebus.CheckHandleSumInShard:
				base := a.Expected.AsBase()
				expectedByResultCell[a.Cell] = base.Uint64()
			}
		}
	}

	for _, round := range sys.Rounds {
		for _, action := range round.VerifierActions {
			fpc, ok := action.(*grandproduct.FinalProductCheck)
			if !ok {
				continue
			}

			resultSlot := fpc.GrandProduct.Result.Context.ID.Slot()
			if err := checkNotLastSlot("result", fpc.GrandProduct.Context().Path(), resultSlot, lastSlot); err != nil {
				return GrandProductSystem{}, err
			}

			query := GrandProductQuery{
				SourceName: fpc.GrandProduct.Context().Path(),
				ResultRef:  ScalarCellRef{Round: resultSlot, Index: fpc.GrandProduct.Result.Context.ID.Position()},
				ZFinalRefs: make([]ScalarCellRef, len(fpc.Entries)),
			}
			for i, e := range fpc.Entries {
				zSlot := e.ZFinal.Context.ID.Slot()
				if err := checkNotLastSlot("z_final", e.ZFinal.Context.Path(), zSlot, lastSlot); err != nil {
					return GrandProductSystem{}, err
				}
				query.ZFinalRefs[i] = ScalarCellRef{Round: zSlot, Index: e.ZFinal.Context.ID.Position()}
			}
			if expected, ok := expectedByResultCell[fpc.GrandProduct.Result]; ok {
				query.HasExpected = true
				query.Expected = expected
			}
			out.Queries = append(out.Queries, query)
		}
	}

	return out, nil
}

File: verifier-ray/src/query/grandproduct.zig
const protocol = @import("../protocol/root.zig");
const base = @import("../field/koalabear.zig");
const ext = @import("../field/koalabear_ext.zig");

pub const Error = error{
    FinalProductMismatch,
    ResultMismatch,
};

// ScalarRef locates a cell in ctx.rounds by its (round, index) coordinates.
// round is the proof.rounds index (0-based); index is the position within that
// round's cells slice. Mirrors logderivativesum.ScalarRef / the vanishing
// sub-verifier's cell_value expression nodes.
pub const ScalarRef = struct {
    round: usize,
    index: usize,
};

// Query is one reduced wiop.GrandProduct. The grandproduct compiler turns it
// into Z running-product columns whose recurrence and row-0 boundary are
// ordinary vanishing constraints — already discharged by the vanishing
// sub-verifier. What remains is:
//
//     ∏_i Z_i[n-1] == Result                        (always)
//     Result == expected.?                           (when expected != null)
//
// The second identity is whichever of prover-ray's grandproduct.CheckResultIsOne
// (permutation arguments; expected == 1) or messagebus.CheckHandleSumInShard
// (message-bus handles; expected is the shard's expected accumulator, one in
// the unsharded case) the compiler attached to this GrandProduct — or neither,
// when a message-bus handle's in-shard check is skipped in favour of a
// downstream cross-shard layer (`expected == null`).
//
// z_final_refs and result_ref are ScalarRefs that index into ctx.rounds at
// verify time, so the verifier reads from the adversary's transcript rather
// than from baked-in honest-prover values.
pub const Query = struct {
    z_final_refs: []const ScalarRef,
    result_ref: ScalarRef,
    expected: ?u64 = null,
};

pub const System = struct {
    queries: []const Query = &.{},
};

pub fn verify(comptime system: System, ctx: protocol.Context) (Error || protocol.CellError)!void {
    inline for (system.queries) |query| {
        // ∏_i Z_i[n-1], reading each Z endpoint from the transcript. `cell` is
        // bounds-checked: the refs are trusted (comptime System) but the proof's
        // round/cells slices are not.
        var prod = ext.Ext.one();
        inline for (query.z_final_refs) |ref| {
            prod = prod.mul((try ctx.cell(ref.round, ref.index)).toExt());
        }

        // The result is also read from the transcript, not baked in.
        const result = (try ctx.cell(query.result_ref.round, query.result_ref.index)).toExt();

        // The final-product identity links the Z endpoints to the claimed result.
        if (!prod.eql(result)) return error.FinalProductMismatch;

        // Whichever boundary predicate the compiler attached — permutation's
        // Result == 1 or message-bus's Result == expected — reduces to this
        // single equality against a compile-time constant.
        if (query.expected) |e| {
            const want = ext.Ext.lift(base.Element.init(e));
            if (!result.eql(want)) return error.ResultMismatch;
        }
    }
}
